src/inmanta/server/agentmanager.py

```diff
@@ -411,10 +411,10 @@ async def _log_session_seen_to_db(
         now = datetime.now().astimezone()
         async with data.AgentProcess.get_connection() as connection:
             async with connection.transaction():
+                await data.AgentProcess.update_last_seen(session.id, now, connection)
```

Comment:
If I understand everything correctly. The locking order is: `environment->agent->agentprocess->agentinstance`. Do you agree? In that case we should move the last line to the top as well.

Reply:
agentprocess->agentinstance -> agent 

is what is used in the rest of the agentmanager.

For the cascading delete, I don't know.

This at least makes it consistent with the rest of the file

Reply:
Hmm. In the `Environment.delete_cascade()` method, the `Agent` is delete first and only then the `AgentProcess`. I am not convinced the locking order is right now, but maybe it's better to move forward with little steps and to see what happens.

Reply:
In the first part of the method: yes, but no transaction there. 

The second part: I don;t know.

Reply:
Indeed. I completely missed the fact that it doesn't run in a transaction.